DM-47980: Add an option to include dimension records into general query result

andy-slac opened the pull request:
The `GeneralQueryResults.iter_tuples` method returned DataIds without dimension records. In some cases (e.g. for obscore export) it would be very useful to include records in the same result to avoid querying them separately. New method `with_dimension_records` is added to the class to trigger adding fields from all dimension records into returned page. This will produce many duplicates for some dimensions (e.g. `instrument`) but it keeps page structure simple.

This adds one attribute to the `GeneralResultSpec` class, will need some care with Butler server compatibility.

## Checklist

- [ ] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

andy-slac committed:
Add an option to include dimension records into general query result (DM-47980)

The `GeneralQueryResults.iter_tuples` method returned DataIds without
dimension records. In some cases (e.g. for obscore export) it would be
very useful to include records in the same result to avoid querying
them separately. New method `with_dimension_records` is added to the class
to trigger adding fields from all dimension records into returned page.
This will produce many duplicates for some dimensions (e.g. `instrument`)
but it keeps page structure simple.

This adds one attribute to the `GeneralResultSpec` class, will need some
care with Butler server compatibility.

TallJimbo commented on python/lsst/daf/butler/direct_query_driver/_result_page_converter.py:
```diff
@@ -326,17 +326,35 @@ class GeneralResultPageConverter(ResultPageConverter):  # numpydoc ignore=PR01
+        # In case `spec.include_dimension_records` is True then in addition to
+        # columns returned by the query we have to add columns from dimension
+        # records that are not returned by the query. These columns belong to
+        # either cached or skypix dimensions.
```
Why do you need to add the cached/skypix columns to the list of column converters?  I'd have thought you could just load entire records from the cache (or construct them on-the fly for skypix) when you expand the data IDs.  That is what the `DataCoordinate` results converter does, and by extension the `DatasetRef` one as well.

Is there a need to have the cached/skypix dimension record fields accessible outside of `iter_tuples`?

andy-slac replied:
My idea was that when a client explicitly asks for dimension records then all their columns for all dimensions need to be present in the returned rows.  Handling it differently between cached and non-cached dimensions will probably complicate client code when that code wants to look at individual fields (as opposed to the DatasetRefs returned from iter_tuples). Potentially we can avoid adding those columns to the GeneralResultPage, and make GeneralResult to add them during iteration, but that would require additional round trips for filling caches in case of remote butler.

TallJimbo commented on python/lsst/daf/butler/queries/_general_query_results.py:
```diff
@@ -125,14 +125,10 @@ def iter_tuples(self, *dataset_types: DatasetType) -> Iterator[GeneralResultTupl
             for dataset_type, dimensions, id_key, run_key in dataset_keys:
-                values = tuple(row[key] for key in itertools.chain(dimensions.required, dimensions.implied))
-                data_id = DataCoordinate.from_full_values(dimensions, values)
+                data_id = self._make_data_id(row, dimensions)
```
Instead of making a brand new data ID, it should be a bit more efficient (especially in memory use) to do:
```
data_id = data_coordinate.subset(dimensions)
```
since we know we've already got a data ID with a superset of these dimensions.

TallJimbo commented on python/lsst/daf/butler/remote_butler/server/handlers/_query_serialization.py:
```diff
@@ -79,11 +79,12 @@ def convert_query_page(spec: ResultSpec, page: ResultPage) -> QueryExecuteResult
+        for row in page.rows
     ]
-    return GeneralResultModel(rows=rows)
+    return GeneralResultModel(rows=rows, columns=[str(column) for column in columns])
```
Oh, are you getting all result columns (including the cached/skypix ones) because you don't want/expect the `RemoteButler` client to have those caches populated?  I think we do...that's a lot of network traffic we can just avoid, and the dimension record cache is the least problematic one we've got in terms of how to manage expiration.

andy-slac replied:
Indeed, I though avoiding the need for client-side cache would be nice 🙂 But it definitely increases data size. Let me think about it again, I'll probably revert to getting those records from cache.

andy-slac replied:
OK, I remember the reason now why I did not want to use cached dimension records inside `GeneralResult` - the cache does not exist at that level. `DimensionRecordCache` is instantiated now only for direct butler (SqlRegistry more specifically) and is passed to `DirectQueryDriver`. Remote butler does not make `DimensionRecordCache` and remote query driver does not do it either. I guess none of the result classes so far needed that cache. @dhirving, what are your thoughts on that, should we add dimension records cache to remote butler?

dhirving replied:
I prefer not to have that cache on the client side.  Even if the entire contents of the cache had to be sent on every request, it is less than 4k gzipped.  So we're not exactly talking about a ton of data.  To me the added complexity of maintaining the cache and embedding the concept of governor dimensions into the client/server protocol doesn't seem worth that small of a savings.  

It's not so bad having that cache on our server because we can restart the server if it changes, but I think we'd want some way to invalidate it if we had it on the client side.  If we didn't, we'd have to track down and restart every service using the client too.  This wouldn't be that hard -- we could send a hash with every response that implies a particular set of cached dimension records.  But there are lots of ways to mess that up, so why open that can of worms to save (at worst) 4k.

For the other result specs, my plan was to use the containers from DM-41114 to pull out the repeated dimension records.  I think the governor dimensions could be treated exactly the same as other dimension records within those containers, and the client side doesn't need to know whether they came from the cache or the DB.  Not sure how well that will work in the context of GeneralResultSpec, though.

TallJimbo replied:
The potential savings here is not the same as the size of the cache - in the cache, the data is normalized, while pulling the values from the cached dimension records out into the main rows duplicates them for each row.  And that can be a really big factor, like repeating an instrument or filter name `n_detectors * n_visits` times.

I may still be assuming the cache is more important than it is, because my intuition on this is based on stuff in QG serialization that isn't quite like-for-like.

TallJimbo replied:
Ah, and David already mentioned the possibility of sending the cache down on every request - so maybe we should just do that, at least for now.

dhirving replied:
Right, that's why I mentioned the container classes that normalize the records.  It gives you most of the benefit without the need for an explicit cache (or in some cases a greater benefit, since it would also help with non-governor dimensions).  Not clear how that would fit in with GeneralResultSpec where you are basically trying to ship a lightly transformed version of the raw database row, though.

In my opinion it wouldn't be wise to go too wild optimizing this right now.  My guess would be we're going to see much more of a slowdown in SIA from the region queries than we are from serialization overhead, though a profile might prove me very wrong.

dhirving replied:
I guess to give a more complete top-down answer:

This feature is being added to support an experimental service which is not planned for public release in DP1 -- and which may not survive the experiment phase.  Given that it's an experiment, doing the simplest thing that could possibly work is reasonable -- and I think that's what Andy's existing code does.

Any speedup from the dimension cache is probably linear with the number of results, so there's not an obvious scalability benefit to the cache.  And in any case this is orders of magnitude faster than the previous iteration, which downloaded half the database on every query.

I also think that once this merges down and we get to the next step, it is likely to reveal additional problems that may end up on Andy's plate.  So I personally would wait a minute before spending time optimizing this.  

We might want to wait on making the new API public to avoid needing to think much about backwards compatibility, though.  It is very easy to ensure a simultaneous release of SIA and Butler server since they're deployed through the same system and Stelios' desk is 20 feet from mine.

TallJimbo replied:
I'm fine with punting any optimization of this and going with the code as is, especially if we can try to avoid establishing an interface that limits our options in the future.

TallJimbo reviewed (approved):
I think you've made this more difficult and less memory-efficient than it needed to be by making the cached/skypix columns appear like regular result columns (a couple of line comments on that subject), but it's possible I just didn't see the motivation.  Looks good aside from that question.

andy-slac committed:
Use  more efficient dimension records representation.

GeneralResultPage now stores cached dimension records in a separate structure
which avoids massive duplication of the records.
